gr-blocks/lib/correctiq_auto_impl.cc

```diff
@@ -0,0 +1,179 @@
+
+    message_port_register_in(pmt::mp("rsync"));
+    set_msg_handler(pmt::mp("rsync"),
+                    [this](pmt::pmt_t msg) { this->handle_resync(msg); });
+
+    message_port_register_out(pmt::mp("sync_start"));
+    message_port_register_out(pmt::mp("offsets"));
+}
+
+void correctiq_auto_impl::send_sync_values()
+{
+    pmt::pmt_t meta = pmt::make_dict();
+
+    meta = pmt::dict_add(meta, pmt::mp("real"), pmt::mp(avg_real));
+    meta = pmt::dict_add(meta, pmt::mp("imag"), pmt::mp(avg_img));
+
+    pmt::pmt_t pdu = pmt::cons(meta, pmt::PMT_NIL);
+    message_port_pub(pmt::mp("offsets"), pdu);
+}
+
+void correctiq_auto_impl::send_syncing()
+{
+    pmt::pmt_t pdu = pmt::cons(pmt::intern("syncing"), pmt::from_bool(true));
+
+    message_port_pub(pmt::mp("sync_start"), pdu);
+}
+
+void correctiq_auto_impl::handle_resync(pmt::pmt_t msg)
+{
+    gr::thread::scoped_lock guard(d_mutex);
+
+    GR_LOG_INFO(d_logger, "Auto Synchronizing...");
+    synchronized = false;
+    sync_counter = 0;
+
+    // reset counters on an unsync
+    avg_real = 0.0;
+    avg_img = 0.0;
+
+    send_syncing();
+}
+
+/*
+ * Our virtual destructor.
+ */
+correctiq_auto_impl::~correctiq_auto_impl() {}
+
+double correctiq_auto_impl::get_freq() { return d_freq; }
+
+float correctiq_auto_impl::get_gain() { return d_gain; }
+
+void correctiq_auto_impl::set_freq(double new_value)
+{
+    gr::thread::scoped_lock guard(d_mutex);
+
+    GR_LOG_INFO(d_logger, "Auto Synchronizing...");
+    d_freq = new_value;
+    synchronized = false;
+    sync_counter = 0;
+
+    // reset counters on an unsync
+    avg_real = 0.0;
+    avg_img = 0.0;
+
+    send_syncing();
+}
+
+void correctiq_auto_impl::set_gain(float new_value)
+{
+    gr::thread::scoped_lock guard(d_mutex);
+
+    GR_LOG_INFO(d_logger, "Auto Synchronizing...");
+
+    d_gain = new_value;
+    synchronized = false;
+    sync_counter = 0;
+
+    // reset counters on an unsync
+    avg_real = 0.0;
+    avg_img = 0.0;
+
+    send_syncing();
+}
+
+int correctiq_auto_impl::work(int noutput_items,
+                              gr_vector_const_void_star& input_items,
+                              gr_vector_void_star& output_items)
+{
+    gr::thread::scoped_lock guard(d_mutex);
+
+    const gr_complex* in = (const gr_complex*)input_items[0];
+    gr_complex* out = (gr_complex*)output_items[0];
+
+    for (int i = 0; i < noutput_items; i++) {
+        if (synchronized) {
+            out[i].real(in[i].real() - avg_real);
+            out[i].imag(in[i].imag() - avg_img);
+        } else {
+            // Synchronizing.  Behave just like normal correctiq.
+            avg_real = ratio * (in[i].real() - avg_real) + avg_real;
+            avg_img = ratio * (in[i].imag() - avg_img) + avg_img;
+
+            out[i].real(in[i].real() - avg_real);
+            out[i].imag(in[i].imag() - avg_img);
+
+            sync_counter++;
+        }
+    }
+
+    if (!synchronized && (sync_counter >= d_max_syncSamples)) {
```

Comment:
Do you want this check to be above where `sync_counter++` happens? Seems like if the "max" number of sync Samples is set, we don't want to go beyond it .. or maybe that's OK?